jerry-core/ecma/builtin-objects/ecma-builtin-array-prototype.c

```diff
@@ -245,40 +246,40 @@ ecma_builtin_array_prototype_object_concat (const ecma_value_t args[], /**< argu
   ecma_object_t *new_array_p = ecma_get_object_from_value (new_array);
   uint32_t new_length = 0;
 
-  ecma_value_t ret_value = ECMA_VALUE_EMPTY;
-
   /* 5.b - 5.c for this_arg */
-  ECMA_TRY_CATCH (concat_this_value,
-                  ecma_builtin_helper_array_concat_value (new_array_p, &new_length, ecma_make_object_value (obj_p)),
-                  ret_value);
-
-  /* 5. */
-  for (uint32_t arg_index = 0;
-       arg_index < args_number && ecma_is_value_empty (ret_value);
-       arg_index++)
+  ecma_value_t concat_this_value = ecma_builtin_helper_array_concat_value (new_array_p,
+                                                                           &new_length,
+                                                                           ecma_make_object_value (obj_p));
+  if (ECMA_IS_VALUE_ERROR (concat_this_value))
   {
-    ECMA_TRY_CATCH (concat_value,
-                    ecma_builtin_helper_array_concat_value (new_array_p, &new_length, args[arg_index]),
-                    ret_value);
-    ECMA_FINALIZE (concat_value);
+    ecma_deref_object (new_array_p);
+    return concat_this_value;
   }
 
-  ECMA_FINALIZE (concat_this_value);
+  JERRY_ASSERT (ecma_is_value_empty (concat_this_value));
 
-  if (ecma_is_value_empty (ret_value))
+  /* 5. */
+  for (uint32_t arg_index = 0; arg_index < args_number; arg_index++)
   {
-    ECMA_TRY_CATCH (set_length_value,
-                    ecma_builtin_array_prototype_helper_set_length (new_array_p, ((ecma_number_t) new_length)),
-                    ret_value);
-    ret_value = new_array;
-    ECMA_FINALIZE (set_length_value);
+    ecma_value_t concat_value = ecma_builtin_helper_array_concat_value (new_array_p, &new_length, args[arg_index]);
+
+    if (ECMA_IS_VALUE_ERROR (concat_value))
+    {
+      ecma_deref_object (new_array_p);
+      return concat_value;
+    }
+
+    JERRY_ASSERT (ecma_is_value_empty (concat_value));
   }
-  else
+
+  ecma_value_t set_length_value = ecma_builtin_array_prototype_helper_set_length (new_array_p,
+                                                                                  ((ecma_number_t) new_length));
+  if (ECMA_IS_VALUE_ERROR (set_length_value))
   {
-    ecma_free_value (new_array);
+    return set_length_value;
```

Comment:
Freeing the `new_array_p` in case of error is not required?

Reply:
Thanks! Good catch.

Reply:
Is this also means that there is no test for this? :)

Reply:
Indeed we do not have test for it. The reason is that to trigger this branch we need to achieve that the newly constructed array's length should be greater than UINT32_MAX. We can create this kind of test but the routine iterates through the whole property chain to find the current indexed property which takes serious amount of time. So IMO we can assume that this deref is necessary but I'd not enclose a test case for this corner case.